File: cost/iona/orbacus/java/main/TestIntfFixedDSI_impl.java
// **********************************************************************
//
// Copyright (c) 2001
// IONA Technologies, Inc.
// Waltham, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

package ORBTest;

import org.omg.CORBA.*;

class TestIntfFixedDSI_impl
    extends org.omg.PortableServer.DynamicImplementation
{
    private ORB m_orb;
    private ORBTest_Fixed.Intf m_ti;

    TestIntfFixedDSI_impl(
	ORB orb,
	ORBTest_Fixed.Intf ti
    )
    {
	m_orb = orb;
	m_ti = ti;
    }

    static final String[] m_ids =
    {
        "IDL:ORBTest_Fixed/Intf:1.0"
    };

    public String[]
    _all_interfaces(
	org.omg.PortableServer.POA poa,
	byte[] object_id
    )
    {
        return m_ids;
    }

    public void
    invoke(
	ServerRequest request
    )
    {
	String name = request.operation();

	boolean ex;
	if (name.length() > 2 && name.endsWith("Ex"))
	{
	    name = name.substring(0, name.length() - 2);
	    ex = true;
	}
	else
	{
	    ex = false;
	}

	if(name.equals("_get_attrFixed"))
	{
	    NVList list = m_orb.create_list(0);
	    request.arguments(list);

	    java.math.BigDecimal ret = m_ti.attrFixed();

	    Any any = m_orb.create_any();
	    any.insert_fixed(ret, m_orb.create_fixed_tc((short)24, (short)8));
	    request.set_result(any);
	    
	    return;
	}

	if(name.equals("_set_attrFixed"))
	{
	    NVList list = m_orb.create_list(0);
	    Any any = m_orb.create_any();
	    any.type(m_orb.create_fixed_tc((short)24, (short)8));
	    list.add_value("", any, org.omg.CORBA.ARG_IN.value);
	    request.arguments(list);

	    java.math.BigDecimal arg = any.extract_fixed();
	    m_ti.attrFixed(arg);
	    
	    return;
	}

	if(name.equals("opFixed"))
	{
	    NVList list = m_orb.create_list(0);
            TypeCode tcFixed = m_orb.create_fixed_tc((short)24, (short)8);
	    Any any0 = m_orb.create_any();
	    Any any1 = m_orb.create_any();
	    Any any2 = m_orb.create_any();
	    any0.type(tcFixed);
	    any1.type(tcFixed);
	    list.add_value("", any0, org.omg.CORBA.ARG_IN.value);
	    list.add_value("", any1, org.omg.CORBA.ARG_INOUT.value);
	    list.add_value("", any2, org.omg.CORBA.ARG_OUT.value);
	    request.arguments(list);

	    java.math.BigDecimal arg0 = any0.extract_fixed();
	    FixedHolder arg1 = new FixedHolder();
	    arg1.value = any1.extract_fixed();
	    FixedHolder arg2 = new FixedHolder();

	    java.math.BigDecimal ret = m_ti.opFixed(arg0, arg1, arg2);

	    if (ex)
	    {
		Any exAny = m_orb.create_any();
		ORBTest_Fixed.ExFixedHelper.insert(
		    exAny, new ORBTest_Fixed.ExFixed(ret)
		);
		request.set_exception(exAny);
	    }
	    else
	    {
		Any result = m_orb.create_any();
		result.insert_fixed(ret, tcFixed);
		request.set_result(result);
		any1.insert_fixed(arg1.value, tcFixed);
		any2.insert_fixed(arg2.value, tcFixed);
	    }
	
	    return;
	}
	System.err.println("DSI implementation: unknown operation: " + name);

        NVList list = m_orb.create_list(0);
        request.arguments(list);

        Any exAny = m_orb.create_any();
        BAD_OPERATIONHelper.insert(exAny, new BAD_OPERATION());
        request.set_exception(exAny);
    }
}

File: cost/iona/orbacus/cpp/main/TestIntfFixedDSI_impl.cpp
// **********************************************************************
//
// Copyright (c) 2001
// IONA Technologies, Inc.
// Waltham, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

#ifndef HAVE_NO_FIXED_TYPE

#include <includes.h>

#include <TestIntfFixedDSI_impl.h>

using namespace CORBA;

TestIntfFixedDSI_impl::TestIntfFixedDSI_impl(
    ORB_ptr orb,
    ORBTest_Fixed::Intf_ptr ti
)
    : m_orb(ORB::_duplicate(orb)),
      m_ti(ORBTest_Fixed::Intf::_duplicate(ti))
{
}

RepositoryId
TestIntfFixedDSI_impl::_primary_interface(
    const PortableServer::ObjectId& oid,
    PortableServer::POA_ptr poa
)
{
    return string_dup("IDL:ORBTest_Fixed/Intf:1.0");
}

CORBA::Boolean
TestIntfFixedDSI_impl::_is_a(
    const char* name
)
    throw (CORBA::SystemException)
{
    if (strcmp(name, "IDL:ORBTest_Fixed/Intf:1.0") == 0)
    {
	return true;
    }
    else
    {
	return PortableServer::ServantBase::_is_a(name);
    }
}

void
TestIntfFixedDSI_impl::invoke(
    ServerRequest_ptr request
)
    throw ()
{
    String_var name = request->operation();

    bool ex;
    if (
	strlen(name) > 2
	&& strcmp((const char*)name + strlen(name) - 2, "Ex") == 0
    )
    {
	name[strlen(name) - 2] = '\0';
	ex = true;
    }
    else
    {
	ex = false;
    }

    if (strcmp(name, "_get_attrFixed") == 0)
    {
	NVList_ptr list;
	m_orb->create_list(0, list);
	request->arguments(list);

	ORBTest_Fixed::TestFixed ret = m_ti->attrFixed();

	Any any;
	any <<= Any::from_fixed(ret, 24, 8);
	request->set_result(any);

	return;
    }

    if (strcmp(name, "_set_attrFixed") == 0)
    {
	NVList_ptr list;
	Any* any;

	m_orb->create_list(0, list);
	TypeCode_var tcFixed = m_orb->create_fixed_tc(24, 8);
	list->add(ARG_IN)->value()->replace(tcFixed, 0);
	request->arguments(list);

	any = list->item(0)->value();
	ORBTest_Fixed::TestFixed arg;
	*any >>= Any::to_fixed(arg, 24, 8);
	m_ti->attrFixed(arg);

	return;
    }

    if (strcmp(name, "opFixed") == 0)
    {
	NVList_ptr list;
	Any* any;

	m_orb->create_list(0, list);
	TypeCode_var tcFixed = m_orb->create_fixed_tc(24, 8);
	list->add(ARG_IN)->value()->replace(tcFixed, 0);
	list->add(ARG_INOUT)->value()->replace(tcFixed, 0);
	list->add(ARG_OUT);
	request->arguments(list);

	any = list->item(0)->value();
	ORBTest_Fixed::TestFixed arg0;
	*any >>= Any::to_fixed(arg0, 24, 8);

	any = list->item(1)->value();
	ORBTest_Fixed::TestFixed arg1;
	*any >>= Any::to_fixed(arg1, 24, 8);

	ORBTest_Fixed::TestFixed arg2;

	ORBTest_Fixed::TestFixed ret = (
	    m_ti->opFixed(arg0, arg1, arg2)
	);

	any = list->item(1)->value();
	*any <<= Any::from_fixed(arg1, 24, 8);

	any = list->item(2)->value();
	*any <<= Any::from_fixed(arg2, 24, 8);

	if (ex)
	{
	    Any exAny;
	    exAny <<= new ORBTest_Fixed::ExFixed(ret);
	    request->set_exception(exAny);
	}
	else
	{
	    Any result;
	    result <<= Any::from_fixed(ret, 24, 8);
	    request->set_result(result);
	}

	return;
    }

}

#endif
